File: backend/app/models/mahjong.py
from enum import Enum
from typing import List, Optional, Dict, Set, Any
from pydantic import BaseModel, Field
from datetime import datetime


class TileType(str, Enum):
    """麻将牌类型"""
    WAN = "wan"    # 万
    TIAO = "tiao"  # 条
    TONG = "tong"  # 筒


class MeldType(str, Enum):
    """面子类型"""
    PENG = "peng"  # 碰
    GANG = "gang"  # 杠
    CHI = "chi"    # 吃


class GangType(str, Enum):
    """杠牌类型"""
    MING_GANG = "ming_gang"  # 明杠（直杠）
    AN_GANG = "an_gang"      # 暗杠
    JIA_GANG = "jia_gang"    # 加杠


class Tile(BaseModel):
    """麻将牌"""
    type: TileType
    value: int = Field(..., ge=1, le=9)  # 1-9
    id: Optional[str] = None  # 牌的唯一标识
    
    def __str__(self) -> str:
        """转换为字符串表示"""
        type_map = {
            TileType.WAN: "万",
            TileType.TIAO: "条",
            TileType.TONG: "筒"
        }
        return f"{self.value}{type_map[self.type]}"
    
    @classmethod
    def from_code(cls, code: int) -> "Tile":
        """从数字编码创建麻将牌"""
        if 1 <= code <= 9:
            return cls(type=TileType.WAN, value=code)
        elif 11 <= code <= 19:
            return cls(type=TileType.TIAO, value=code - 10)
        elif 21 <= code <= 29:
            return cls(type=TileType.TONG, value=code - 20)
        else:
            raise ValueError(f"Invalid tile code: {code}")
    
    def to_code(self) -> int:
        """转换为数字编码"""
        if self.type == TileType.WAN:
            return self.value
        elif self.type == TileType.TIAO:
            return self.value + 10
        elif self.type == TileType.TONG:
            return self.value + 20
        else:
            raise ValueError(f"Invalid tile type: {self.type}")


class Meld(BaseModel):
    """面子"""
    id: Optional[str] = None  # 面子的唯一标识
    type: MeldType
    tiles: List[Tile]
    exposed: bool = True
    gang_type: Optional[GangType] = None  # 杠牌类型
    source_player: Optional[int] = None  # 来源玩家ID
    original_peng_id: Optional[str] = None  # 加杠时原碰牌ID
    timestamp: Optional[float] = Field(default_factory=lambda: datetime.now().timestamp())


class HandTiles(BaseModel):
    """玩家手牌"""
    tiles: Optional[List[Tile]] = None  # 允许为None（其他玩家的手牌）
    tile_count: Optional[int] = 0  # 手牌数量（用于其他玩家）
    melds: List[Meld] = []
    # 胜利状态相关字段
    is_winner: Optional[bool] = None
    win_type: Optional[str] = None  # "zimo"自摸 或 "dianpao"点炮
    win_tile: Optional[Tile] = None
    dianpao_player_id: Optional[int] = None  # 点炮玩家ID
    missing_suit: Optional[str] = None  # 定缺花色
    
    class Config:
        # 允许额外字段，保持向后兼容
        extra = "allow"


class PlayerAction(BaseModel):
    """玩家动作"""
    player_id: int
    action_type: Optional[str] = None  # 某些历史记录可能使用 'type' 字段
    type: Optional[str] = None  # 兼容旧格式
    tiles: Optional[List[Tile]] = []  # 改为可选，某些动作可能没有
    tile: Optional[Tile] = None  # 单个牌的动作
    source_player: Optional[int] = None  # 来源玩家
    source_player_id: Optional[int] = None  # 兼容字段
    timestamp: Optional[float] = Field(default_factory=lambda: datetime.now().timestamp())
    
    class Config:
        # 允许额外字段，保持向后兼容
        extra = "allow"


class GameState(BaseModel):
    """游戏状态"""
    game_id: str
    player_hands: Dict[str, HandTiles] = {}
    current_player: int = 0
    discarded_tiles: List[Tile] = []
    player_discarded_tiles: Dict[str, List[Tile]] = {}
    actions_history: List[PlayerAction] = []
    game_started: bool = False
    last_action: Optional[Dict[str, Any]] = None
    show_all_hands: Optional[bool] = None  # 是否显示所有玩家手牌
    
    class Config:
        # 允许额外字段，保持向后兼容
        extra = "allow"
    
    def get_visible_tiles(self) -> List[Tile]:
        """获取所有可见的牌"""
        visible = []
        visible.extend(self.discarded_tiles)
        
        for hand in self.player_hands.values():
            for meld in hand.melds:
                if meld.exposed:
                    visible.extend(meld.tiles)
        
        return visible
    
    def calculate_remaining_tiles(self) -> int:
        """计算剩余牌数（包括所有已使用的牌）"""
        total_tiles = 108  # 标准麻将总牌数
        used_tiles = 0
        
        # 计算所有玩家手牌数量
        for hand in self.player_hands.values():
            # 使用tile_count字段，如果不存在则使用tiles长度
            if hand.tile_count is not None:
                used_tiles += hand.tile_count
            elif hand.tiles is not None:
                used_tiles += len(hand.tiles)
            
            # 计算碰牌杠牌数量
            for meld in hand.melds:
                used_tiles += len(meld.tiles)
        
        # 计算弃牌数量
        used_tiles += len(self.discarded_tiles)
        
        return max(0, total_tiles - used_tiles)

    def calculate_visible_remaining_tiles(self) -> int:
        """计算基于可见牌的剩余牌数"""
        total_tiles = 108
        used_tiles = 0
        
        # 计算所有玩家手牌数量
        for player_id, hand in self.player_hands.items():
            # 只计算"我"（player_id=0）的手牌
            if player_id == "0":
                if hand.tile_count is not None:
                    used_tiles += hand.tile_count
                elif hand.tiles is not None:
                    used_tiles += len(hand.tiles)
            
            # 计算碰牌杠牌数量
            for meld in hand.melds:
                if meld.type == MeldType.GANG and meld.gang_type == GangType.AN_GANG:
                    # 暗杠：只计算"我"的暗杠
                    if player_id == "0":
                        used_tiles += len(meld.tiles)
                else:
                    # 明牌（碰牌、明杠）：所有玩家的都要计算
                    used_tiles += len(meld.tiles)
        
        # 计算弃牌数量 - 所有玩家的弃牌都是可见的
        used_tiles += len(self.discarded_tiles)
        
        return max(0, total_tiles - used_tiles)

    def calculate_remaining_tiles_by_type(self) -> Dict[str, int]:
        """计算每种牌的剩余数量（基于可见牌）"""
        # 初始化每种牌的数量为4张
        remaining_counts = {}
        
        # 初始化万、条、筒 1-9 各4张
        for tile_type in [TileType.WAN, TileType.TIAO, TileType.TONG]:
            for value in range(1, 10):
                remaining_counts[f"{tile_type}-{value}"] = 4
        
        # 收集所有已使用的可见牌
        used_tiles = []
        
        # 收集玩家的手牌和碰杠牌
        for player_id, hand in self.player_hands.items():
            # 只收集"我"的手牌
            if player_id == "0" and hand.tiles is not None:
                used_tiles.extend(hand.tiles)
            
            # 收集所有玩家的碰杠牌（除了暗杠）
            for meld in hand.melds:
                if meld.type == MeldType.GANG and meld.gang_type == GangType.AN_GANG:
                    if player_id == "0":  # 只收集"我"的暗杠
                        used_tiles.extend(meld.tiles)
                else:  # 收集所有明牌
                    used_tiles.extend(meld.tiles)
        
        # 收集弃牌
        used_tiles.extend(self.discarded_tiles)
        
        # 更新剩余数量
        for tile in used_tiles:
            key = f"{tile.type}-{tile.value}"
            if key in remaining_counts:
                remaining_counts[key] = max(0, remaining_counts[key] - 1)
        
        return remaining_counts


class AnalysisResult(BaseModel):
    """分析结果"""
    recommended_discard: Optional[Tile] = None
    discard_scores: Dict[str, float] = {}
    listen_tiles: List[Tile] = []
    win_probability: float = 0.0
    remaining_tiles_count: Dict[int, int] = {}
    suggestions: List[str] = []


class GameRequest(BaseModel):
    """游戏请求"""
    game_state: GameState
    target_player: int = 0  # 目标玩家ID（通常是自己）


class GameResponse(BaseModel):
    """游戏响应"""
    success: bool
    analysis: Optional[AnalysisResult] = None
    message: str = "" 


# 操作相关模型
class TileOperationRequest(BaseModel):
    """牌操作请求"""
    player_id: int = Field(..., ge=0, le=3)
    operation_type: str  # discard, peng, gang, hand
    tile: Tile
    source_player_id: Optional[int] = None
    game_id: Optional[str] = None  # 添加游戏ID字段


class GameOperationResponse(BaseModel):
    """游戏操作响应"""
    success: bool
    message: str
    game_state: Optional[Dict] = None


class GameStateRequest(BaseModel):
    """游戏状态请求"""
    game_state: GameState


class ResetGameResponse(BaseModel):
    """重置游戏响应"""
    success: bool
    message: str

File: frontend/src/types/mahjong.ts
export enum TileType {
  WAN = "m",        // 万 - 统一使用m表示
  TIAO = "s",       // 条 - 统一使用s表示  
  TONG = "p",       // 筒 - 统一使用p表示
  ZI = "z"          // 字牌 - 统一使用z表示
}

// 保持兼容性的旧枚举（用于显示）
export enum TileTypeDisplay {
  WAN = "wan",      // 万
  TIAO = "tiao",    // 条
  TONG = "tong",    // 筒
  ZI = "zi"         // 字牌
}

export interface Tile {
  type: TileType;
  value: number;
  id?: string;      // 添加牌的唯一标识
}

export enum MeldType {
  PENG = "peng",    // 碰
  GANG = "gang",    // 杠
  CHI = "chi"       // 吃
}

export enum GangType {
  MING_GANG = "ming_gang",  // 明杠（直杠）
  AN_GANG = "an_gang",      // 暗杠
  JIA_GANG = "jia_gang"     // 加杠
}

export interface Meld {
  id?: string;      // 添加唯一标识
  type: MeldType;
  tiles: Tile[];
  exposed: boolean;
  gang_type?: GangType; // 杠牌类型（仅当type为GANG时有效）
  source_player?: number;        // 直杠时：被杠牌的来源玩家ID
  original_peng_id?: string;     // 加杠时：原碰牌的ID
  timestamp?: number;            // 操作时间戳
}

export interface HandTiles {
  tiles: Tile[] | null;  // 其他玩家为null，只有我（玩家0）有具体牌面
  tile_count: number;     // 手牌数量（必需字段）
  melds: Meld[];
  missing_suit?: string | null;  // 定缺花色：wan/tiao/tong
  is_winner?: boolean;    // 是否胜利
  win_type?: 'zimo' | 'dianpao';  // 胜利类型：自摸或点炮
  win_tile?: Tile;        // 胡牌牌
  dianpao_player_id?: number;  // 点炮者ID（仅点炮胡牌时有效）
}

export interface PlayerAction {
  player_id: number;
  action_type: string;
  tiles: Tile[];
  timestamp?: number;
}

export interface GameState {
  game_id: string;
  player_hands: {
    [key: string]: HandTiles;
  };
  current_player: number;
  discarded_tiles: Tile[];
  player_discarded_tiles: {
    [key: string]: Tile[];
  };
  actions_history: PlayerAction[];
  game_started: boolean;
  game_ended?: boolean;  // 牌局是否结束
  show_all_hands?: boolean;  // 是否显示所有玩家手牌
}

export interface AnalysisResult {
  recommended_discard?: Tile;
  discard_scores: { [key: string]: number };
  listen_tiles: Tile[];
  win_probability: number;
  remaining_tiles_count: { [key: number]: number };
  suggestions: string[];
  ultimate_results?: UltimateAnalysisResult[];  // 血战到底分析结果
}

// 血战到底分析结果接口
export interface UltimateAnalysisResult {
  discard_tile: string;
  expected_value: number;    // 收益
  jinzhang_types: number;    // 进张种类数
  jinzhang_count: number;    // 总进张数
  jinzhang_detail: string;   // 详细进张信息
  shanten: number;           // 向听数
  can_win: boolean;          // 是否可胡
  is_forced: boolean;        // 是否必出
  patterns: string[];        // 特殊牌型
}

export interface GameRequest {
  game_state: GameState;
  target_player: number;
}

export interface GameResponse {
  success: boolean;
  analysis?: AnalysisResult;
  message: string;
}

export interface TileInfo {
  code: number;
  type: string;
  value: number;
  display: string;
}

// 胜利者信息接口
export interface Winner {
  player_id: number;
  win_type: 'zimo' | 'dianpao';
  win_tile?: Tile;
  dianpao_player_id?: number;
}

// 输入模式枚举
export enum InputMode {
  ADD_TO_HAND = "add_to_hand",
  DISCARD = "discard", 
  PENG = "peng",
  MING_GANG = "ming_gang",
  AN_GANG = "an_gang"
}

// 工具函数
export const createTile = (type: TileType, value: number): Tile => ({
  type,
  value
});

export const tileToCode = (tile: Tile): number => {
  switch (tile.type) {
    case TileType.WAN:
      return tile.value;
    case TileType.TIAO:
      return tile.value + 10;
    case TileType.TONG:
      return tile.value + 20;
    case TileType.ZI:
      return tile.value + 30;
    default:
      throw new Error(`Invalid tile type: ${tile.type}`);
  }
};

export const codeToTile = (code: number): Tile => {
  if (1 <= code && code <= 9) {
    return { type: TileType.WAN, value: code };
  } else if (11 <= code && code <= 19) {
    return { type: TileType.TIAO, value: code - 10 };
  } else if (21 <= code && code <= 29) {
    return { type: TileType.TONG, value: code - 20 };
  } else if (31 <= code && code <= 37) {
    return { type: TileType.ZI, value: code - 30 };
  } else {
    throw new Error(`Invalid tile code: ${code}`);
  }
};

export const tileToString = (tile: Tile): string => {
  const typeMap = {
    [TileType.WAN]: "万",
    [TileType.TIAO]: "条", 
    [TileType.TONG]: "筒",
    [TileType.ZI]: ["", "东", "南", "西", "北", "中", "发", "白"][tile.value] || ""
  };
  
  if (tile.type === TileType.ZI) {
    return typeMap[tile.type];
  } else {
    return `${tile.value}${typeMap[tile.type]}`;
  }
};

// 转换为mps格式（用于后端API）
export const tileToMpsString = (tile: Tile): string => {
  return `${tile.value}${tile.type}`;
};

// 从mps格式转换为Tile对象
export const mpsStringToTile = (mpsStr: string): Tile => {
  if (mpsStr.length !== 2) {
    throw new Error(`Invalid mps string: ${mpsStr}`);
  }
  
  const value = parseInt(mpsStr[0]);
  const type = mpsStr[1] as TileType;
  
  if (!Object.values(TileType).includes(type)) {
    throw new Error(`Invalid tile type: ${type}`);
  }
  
  return { type, value };
};

// 转换手牌数组为mps字符串
export const tilesToMpsString = (tiles: Tile[]): string => {
  const grouped: { [key in TileType]?: number[] } = {};
  
  tiles.forEach(tile => {
    if (!grouped[tile.type]) {
      grouped[tile.type] = [];
    }
    grouped[tile.type]!.push(tile.value);
  });
  
  let result = "";
  
  // 按照 m, p, s, z 的顺序
  const order: TileType[] = [TileType.WAN, TileType.TONG, TileType.TIAO, TileType.ZI];
  
  order.forEach(type => {
    if (grouped[type]) {
      const sortedValues = grouped[type]!.sort((a, b) => a - b);
      result += sortedValues.join('') + type;
    }
  });
  
  return result;
};

export const tilesEqual = (tile1: Tile, tile2: Tile): boolean => {
  return tile1.type === tile2.type && tile1.value === tile2.value;
};

// 计算剩余牌数（原有逻辑：基于tile_count计算所有已使用的牌）
export const calculateRemainingTiles = (gameState: GameState): number => {
  const totalTiles = 108; // 标准麻将总牌数
  
  // 计算已使用的牌数
  let usedTiles = 0;
  
  // 计算所有玩家手牌数量
  Object.values(gameState.player_hands).forEach(hand => {
    // 使用tile_count字段，如果不存在则用tiles数组长度（向后兼容）
    const handTileCount = hand.tile_count !== undefined ? hand.tile_count : (hand.tiles?.length || 0);
    usedTiles += handTileCount;
    
    // 计算碰牌杠牌数量（所有碰杠都计算）
    hand.melds.forEach(meld => {
      usedTiles += meld.tiles.length;
    });
  });
  
  // 计算弃牌数量
  usedTiles += gameState.discarded_tiles.length;
  
  return Math.max(0, totalTiles - usedTiles);
};

// 计算基于可见牌的剩余牌数（用于未出牌数计算）
export const calculateVisibleRemainingTiles = (gameState: GameState): number => {
  const totalTiles = 108; // 标准麻将总牌数
  
  // 计算已使用的可见牌数
  let usedTiles = 0;
  
  // 计算所有玩家手牌数量
  Object.entries(gameState.player_hands).forEach(([playerIdStr, hand]) => {
    const playerId = parseInt(playerIdStr);
    
    // 计算手牌：游戏结束时所有玩家手牌可见，平时只有"我"的手牌可见
    if (playerId === 0 || gameState.game_ended) {
      const handTileCount = hand.tile_count !== undefined ? hand.tile_count : (hand.tiles?.length || 0);
      usedTiles += handTileCount;
    }
    
    // 计算碰牌杠牌数量
    hand.melds.forEach(meld => {
      if (meld.type === MeldType.GANG && meld.gang_type === GangType.AN_GANG) {
        // 暗杠：游戏结束时所有玩家暗杠可见，平时只有"我"的暗杠可见
        if (playerId === 0 || gameState.game_ended) {
          usedTiles += meld.tiles.length;
        }
      } else {
        // 明牌（碰牌、明杠）：所有玩家的都要计算
        usedTiles += meld.tiles.length;
      }
    });
  });
  
  // 计算弃牌数量 - 所有玩家的弃牌都是可见的
  usedTiles += gameState.discarded_tiles.length;
  
  return Math.max(0, totalTiles - usedTiles);
};

// 计算每种牌的剩余数量（基于可见牌）
export const calculateRemainingTilesByType = (gameState: GameState): { [key: string]: number } => {
  console.log('🔍 开始计算剩余牌数...');
  
  // 初始化每种牌的数量为4张
  const remainingCounts: { [key: string]: number } = {};
  
  // 万子 1-9
  for (let i = 1; i <= 9; i++) {
    remainingCounts[`${TileType.WAN}-${i}`] = 4;
  }
  
  // 条子 1-9  
  for (let i = 1; i <= 9; i++) {
    remainingCounts[`${TileType.TIAO}-${i}`] = 4;
  }
  
  // 筒子 1-9
  for (let i = 1; i <= 9; i++) {
    remainingCounts[`${TileType.TONG}-${i}`] = 4;
  }
  
  console.log('📦 初始牌数:', remainingCounts);
  
  // 收集所有已使用的可见牌
  const usedTiles: Tile[] = [];
  
  // 收集玩家的手牌和碰杠牌
  Object.entries(gameState.player_hands).forEach(([playerIdStr, hand]) => {
    const playerId = parseInt(playerIdStr);
    
    // 获取手牌数量，安全处理null的情况
    const handTileCount = hand.tile_count !== undefined ? hand.tile_count : (hand.tiles?.length || 0);
    const handTiles = hand.tiles || [];
    
    console.log(`🏠 处理玩家${playerId}:`, {
      手牌数量: handTileCount,
      碰杠数量: hand.melds.length,
      具体手牌: playerId === 0 ? handTiles.map(t => `${t.value}${t.type}`) : '(隐藏)'
    });
    
    // 收集手牌：游戏结束时所有玩家手牌可见，平时只有"我"的手牌可见
    if (handTiles.length > 0 && (playerId === 0 || gameState.game_ended)) {
      console.log(`👤 收集玩家${playerId}的手牌:`, handTiles.map(t => `${t.value}${t.type}`));
      usedTiles.push(...handTiles);
    }
    
    // 胜利牌现在已经包含在最终手牌中，不需要单独计算
    
    // 收集碰牌杠牌
    hand.melds.forEach((meld, meldIndex) => {
      console.log(`🎴 处理玩家${playerId}的第${meldIndex}个碰杠:`, {
        类型: meld.type,
        杠牌类型: meld.gang_type,
        是否明牌: meld.exposed,
        牌数量: meld.tiles.length,
        牌内容: meld.tiles.map(t => `${t.value}${t.type}`)
      });
      
      if (meld.type === MeldType.GANG && meld.gang_type === GangType.AN_GANG) {
        // 暗杠：游戏结束时所有玩家暗杠可见，平时只有"我"的暗杠可见
        if (playerId === 0 || gameState.game_ended) {
          console.log(`🔒 收集玩家${playerId}的暗杠牌`);
          usedTiles.push(...meld.tiles);
        } else {
          console.log(`🔒 跳过玩家${playerId}的暗杠牌（游戏未结束）`);
        }
      } else {
        // 明牌（碰牌、明杠）：所有玩家的都要收集
        console.log('👁️ 收集明牌:', meld.tiles.map(t => `${t.value}${t.type}`));
        usedTiles.push(...meld.tiles);
      }
    });
  });
  
  // 收集弃牌 - 所有玩家的弃牌都是可见的
  console.log('🗑️ 收集弃牌:', gameState.discarded_tiles.map(t => `${t.value}${t.type}`));
  usedTiles.push(...gameState.discarded_tiles);
  
  console.log('📊 所有已使用的牌:', usedTiles.map(t => `${t.value}${t.type}`));
  
  // 减去已使用的牌
  usedTiles.forEach(tile => {
    const key = `${tile.type}-${tile.value}`;
    console.log(`🔢 处理牌 ${tile.value}${tile.type}, key: ${key}, 当前剩余: ${remainingCounts[key]}`);
    if (remainingCounts[key] !== undefined) {
      remainingCounts[key] = Math.max(0, remainingCounts[key] - 1);
      console.log(`  ➡️ 减少后剩余: ${remainingCounts[key]}`);
    } else {
      console.warn(`  ⚠️ 未知的牌类型key: ${key}`);
    }
  });
  
  console.log('✅ 最终剩余牌数:', remainingCounts);
  return remainingCounts;
};

// 综合分析相关类型定义
export interface ComprehensiveAnalysisChoice {
  tile: string;           // 打牌选择 (如 "1m")
  number: number;         // 有效牌数
  tiles: string[];        // 有效牌列表
}

export interface ComprehensiveAnalysisResult {
  method: 'tenhou_website' | 'local_simulation' | 'exhaustive';
  method_name: string;
  success: boolean;
  error_message?: string;
  choices: ComprehensiveAnalysisChoice[];
  analysis_time: number;
  timestamp: string;
}

export interface ComprehensiveAnalysisResponse {
  hand: string;
  hand_display: string;
  results: ComprehensiveAnalysisResult[];
  comparison?: {
    success_rate: { [key: string]: boolean };
    performance: { [key: string]: string };
    choice_consistency: {
      match_rate: string;
      percentage: string;
    };
    summary: {
      total_methods: number;
      successful_methods: number;
      fastest_method?: string;
    };
  };
}

export interface ComprehensiveAnalysisRequest {
  hand: string;
  methods: Array<'tenhou_website' | 'local_simulation' | 'exhaustive'>;
  tile_format: 'mps' | 'frontend';
} 